### PR TITLE
CLOUDSTACK-10290: Introduce option to have config drives on primary storage

### PR DESCRIPTION
**Pull request by rhtyd**
This introduces a new global setting `vm.configdrive.primarypool.enabled` to toggle creation/hosting of config drive iso files on primary storage, the default will be `false` causing them to be hosted on secondary storage. The support is limited from hypervisor resource side and in current implementation limited to KVM. The next big change is that config drive is created at a temporary location by management server and shipped to either KVM or SSVM agent via cmd-answer pattern, the data of which is not logged in logs. This saves us from adding `genisoimage` dependency on cloudstack-agent pkg.

The APIs to reset ssh public key, password and user-data (via update VM API) requires that VM should be shutdown. Therefore, in the refactoring I removed the case of updation of existing ISO. If there are objections I'll re-put the strategy to detach+attach new config iso as a way of updation. In the refactored implementation, the folder name is changed to lower-cased `configdrive`. And during VM start, migration or shutdown/removal if primary storage is enable for use, the KVM agent will handle cleanup tasks otherwise SSVM agent will handle them. 

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [x] Cleanup (Code refactoring and cleanup, that may add test cases)

## Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
Testing
- [x] I have added tests to cover my changes.
- [x] All relevant new and existing integration tests have passed.
- [x] A full integration testsuite with all test that can run on my environment has passed.

### CODE REVIEW

**Comment by rhtyd** [comment]
Work left is some testing around VM migration and unit test/marvin test fixing.
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test matrix

**Comment by borisstoyanov** [comment]
@blueorangutan test centos7 xenserver-71

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

File: engine/storage/configdrive/src/org/apache/cloudstack/storage/configdrive/ConfigDriveBuilder.java [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+            File openStackFolder = new File(tempDirName + ConfigDrive.openStackConfigDriveName);
+            if (openStackFolder.exists() || openStackFolder.mkdirs()) {
+                writeFile(openStackFolder, "vendor_data.json", "{}");
+                writeFile(openStackFolder, "network_data.json", "{}");
+            } else {
+                throw new CloudRuntimeException("Failed to create folder " + openStackFolder);
+            }
+
+            JsonObject metaData = new JsonObject();
+            for (String[] item : vmData) {
```
Comment (nvazquez) [inline engine/storage/configdrive/src/org/apache/cloudstack/storage/configdrive/ConfigDriveBuilder.java]: Can this logic in the FOR loop be extracted to a method?

File: engine/storage/configdrive/src/org/apache/cloudstack/storage/configdrive/ConfigDriveBuilder.java [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+            writeFile(openStackFolder, "meta_data.json", metaData.toString());
+
+            String linkResult = linkUserData(tempDirName);
+            if (linkResult != null) {
+                String errMsg = "Unable to create user_data link due to " + linkResult;
+                throw new CloudRuntimeException(errMsg);
+            }
+
+            File tmpIsoStore = new File(tempDirName, new File(isoFileName).getName());
+            Script command = new Script("/usr/bin/genisoimage", Duration.standardSeconds(300), LOG);
```
Comment (nvazquez) [inline engine/storage/configdrive/src/org/apache/cloudstack/storage/configdrive/ConfigDriveBuilder.java]: Can this also be extracted into a method?

File: engine/storage/configdrive/src/org/apache/cloudstack/storage/configdrive/ConfigDriveBuilder.java (hunk 1 above)
Comment (rhtyd) [reply engine/storage/configdrive/src/org/apache/cloudstack/storage/configdrive/ConfigDriveBuilder.java]: To limit my effort scope, code in this class were previously in Nfs resource class. Most of the changes in this class are from a move-refactor. I'll further refactor if I've time, so yes I can try.

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
Merging based on code reviews, test results incl. manual tests.

**Pull request merged**